witness/internal/witness/witness.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"fmt"
+
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/merkle/logverifier"
+	"golang.org/x/mod/sumdb/note"
+)
+
+type Chkpt struct {
+	Parsed *log.Checkpoint
+	Raw    []byte
+}
+
+type ChkptStorage interface {
+	// GetLatest returns the latest checkpoint for a given log.
+	GetLatest(logPK string) (*Chkpt, error)
+
+	// SetCheckpoint adds a checkpoint to the storage for a given log.
+	SetCheckpoint(logPK string, c Chkpt) error
+}
+
+type Witness struct {
+	db     ChkptStorage
+	Signer note.Signer
+	Hasher hashers.LogHasher
+	LogV   logverifier.LogVerifier
+	SigVs  []note.Verifier
+}
+
+// NewWitness creates a new witness, which initially has no logs to follow.
+func NewWitness(h hashers.LogHasher, db ChkptStorage, sk string) (*Witness, error) {
+	ns, err := note.NewSigner(sk)
+	if err != nil {
+		return nil, err
+	}
+	return &Witness{
+		db:     db,
+		Signer: ns,
+		Hasher: h,
+		LogV:   logverifier.New(h),
+		SigVs:  []note.Verifier{},
+	}, nil
+}
+
+// parse verifies the checkpoint under the given list of possible public keys.
+// It returns the parsed checkpoint and the list of public keys under which the
+// signature(s) verified.
+func (w *Witness) parse(chkptRaw []byte, verifiers []note.Verifier) (*log.Checkpoint, []string, error) {
+	n, err := note.Open(chkptRaw, note.VerifierList(verifiers...))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
+	chkpt := &log.Checkpoint{}
+	_, err = chkpt.Unmarshal([]byte(n.Text))
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal new checkpoint: %w", err)
+	}
+	var keys []string
```

Comment:
When you know the size the array will be, it's good practice to allocate it at the correct size. Here you would change this to `keys := make([]string, 0, len(n.Sigs))`. Or more directly (i.e. avoid the append):

```
keys := make([]string, len(n.Sigs))
for i, sig := range n.Sigs {
  keys[i] = sig.Name
}
```

Reply:
Done, thanks!